### PR TITLE
[compiler-rt][fuchsia] Preallocate a vmar for sanitizer internals

### PR DESCRIPTION
**Pull request by PiJoules**
In an effort to reduce more mmap fragmentation, allocate a large enough vmar where we can map sanitizer internals via DoAnonymousMmap. Objects being mapped here include asan's FakeStack, LowLevelAllocator mappings, the primary allocator's TwoLevelMap, InternalMmapVector, StackStore, and asan's thread internals. The vmar is large enough to hold the total size of these objects seen in a "typical" process lifetime. If the vmar is full, it will fallback to mapping in the root vmar.

### CODE REVIEW

**Comment by PiJoules** [comment]
WIP for now and will request a formal review once I assert I can do 100 runs of the boot-libc-unittests on bringup.riscv64-asan without a flake.

**Comment by PiJoules** [comment]
Ok now requesting a formal review since it ran 100 times without flaking. Prior to this, it would flake about every 10 runs.

**Comment by PiJoules** [comment]
*ping*

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp [hunk 1]
```diff
@@ -129,6 +129,55 @@ uptr GetMaxVirtualAddress() { return GetMaxUserVirtualAddress(); }
 
 bool ErrorIsOOM(error_t err) { return err == ZX_ERR_NO_MEMORY; }
 
+// For any sanitizer internal that needs to map something which can be unmmaped
```
Comment (frobtech) [inline compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:132]: typo: "unmapped"

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp [hunk 2]
```diff
@@ -129,6 +129,55 @@ uptr GetMaxVirtualAddress() { return GetMaxUserVirtualAddress(); }
...
+// later, first attempt to map to a pre-allocated vmar. This helps reduce
```
Comment (frobtech) [inline compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:133]: I find it more readable to use caps for acronyms like VMAR.

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp [hunk 3]
```diff
@@ -129,6 +129,55 @@ uptr GetMaxVirtualAddress() { return GetMaxUserVirtualAddress(); }
...
+// fragmentation from many small anonymous mmap calls. A good value for this
+// vmar size would be the total size of your typical sanitizer internal objects
+// allocated in an "average" process lifetime. Examples of this include:
+// FakeStack, LowLevelAllocator mappings, TwoLevelMap, InternalMmapVector,
+// StackStore, CreateAsanThread, etc.
+//
+// 0xc52000 is ~12.32MB. This is roughly equal to the total sum of sanitizer
+// internal mappings in a run of boot-libc-unittests.
```
Comment (frobtech) [inline compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:141]: I don't think we should be referring to individual Fuchsia tree things here.  We can say things more generic like "a large test case".  Let's use a more round number.  If ~12.32MiB is a good number, then 13MiB is a good number, and you can write that as `13 << 20; // 13 MiB` or something more readable like that.

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp [hunk 4]
```diff
@@ -129,6 +129,55 @@ uptr GetMaxVirtualAddress() { return GetMaxUserVirtualAddress(); }
...
+constexpr size_t kSanitizerHeapVmarSize = 0xc52000;
+static zx_handle_t gSanitizerHeapVmar = 0;
```
Comment (frobtech) [inline compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:143]: It's fine to just assume that default zero initialization is right here, but if you're going to be explicit then use `ZX_HANDLE_INVALID`.

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp [hunk 5]
```diff
@@ -129,6 +129,55 @@ uptr GetMaxVirtualAddress() { return GetMaxUserVirtualAddress(); }
...
+
+static zx_status_t GetSanitizerHeapVmar(zx_handle_t *vmar) {
+  zx_status_t status = ZX_OK;
+  if (!gSanitizerHeapVmar) {
```
Comment (frobtech) [inline compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:147]: Don't use implicit bool coercion for non-pointers.  Use `!= ZX_HANDLE_INVALID`.

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp [hunk 6]
```diff
@@ -129,6 +129,55 @@ uptr GetMaxVirtualAddress() { return GetMaxUserVirtualAddress(); }
...
+    CHECK_EQ(kSanitizerHeapVmarSize % GetPageSizeCached(), 0);
+    uintptr_t base;
+    status = _zx_vmar_allocate(
+        _zx_vmar_root_self(),
+        ZX_VM_CAN_MAP_READ | ZX_VM_CAN_MAP_WRITE | ZX_VM_CAN_MAP_SPECIFIC, 0,
+        kSanitizerHeapVmarSize, &gSanitizerHeapVmar, &base);
+  }
+  *vmar = gSanitizerHeapVmar;
+  if (status == ZX_OK)
+    CHECK_NE(gSanitizerHeapVmar, 0);
```
Comment (frobtech) [inline compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:157]: `ZX_HANDLE_INVALID`.

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp [hunk 7]
```diff
@@ -129,6 +129,55 @@ uptr GetMaxVirtualAddress() { return GetMaxUserVirtualAddress(); }
...
+  return status;
+}
+
+static zx_status_t TryVmoMapSanitizerVmar(zx_vm_option_t options,
+                                          size_t vmar_offset, zx_handle_t vmo,
+                                          size_t size, uintptr_t *addr) {
+  zx_status_t status;
+  zx_handle_t vmar;
+  if ((status = GetSanitizerHeapVmar(&vmar)) != ZX_OK)
```
Comment (frobtech) [inline compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:166]: Prefer initializing in the declaration to setting the variable inside the `if`.

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp [hunk 8]
```diff
@@ -243,6 +290,12 @@ void UnmapOrDieVmar(void *addr, uptr size, zx_handle_t target_vmar) {
 
   zx_status_t status =
       _zx_vmar_unmap(target_vmar, reinterpret_cast<uintptr_t>(addr), size);
+  if (status != ZX_OK && target_vmar == gSanitizerHeapVmar) {
```
Comment (frobtech) [inline compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:293]: Check for the specific error code returned for this case, which is `ZX_ERR_INVALID_ARGS`.

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp [hunk 9]
```diff
@@ -357,7 +406,7 @@ void *MmapAlignedOrDieOnFatalError(uptr size, uptr alignment,
 }
 
 void UnmapOrDie(void *addr, uptr size) {
-  UnmapOrDieVmar(addr, size, _zx_vmar_root_self());
+  UnmapOrDieVmar(addr, size, gSanitizerHeapVmar);
```
Comment (frobtech) [inline compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:409]: This needs to use the fallback logic in case the root vmar was used, as above.

**Review by frobtech — CHANGES_REQUESTED** [review]

File: compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp (hunk 9 above)
Comment (frobtech) [reply compiler-rt/lib/sanitizer_common/sanitizer_fuchsia.cpp:409]: I see, thanks. The way I saw the diff that was hard to find, probably just my lack of familiarity with the GH code review UI.

**Review by frobtech — APPROVED** [review]

**Pull request merged**